cpp/include/raft/core/device_buffer.hpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "device_mdarray.hpp"
+#include "device_mdspan.hpp"
+
+#include <raft/util/cudart_utils.hpp>
+
+#include <variant>
+
+namespace raft {
+
+template <typename ElementType,
+          typename Extents,
+          typename LayoutPolicy    = layout_c_contiguous,
+          typename ContainerPolicy = detail::device_uvector_policy<std::remove_cv_t<ElementType>>>
+class device_buffer {
+  using view_type            = device_mdspan<ElementType, Extents, LayoutPolicy>;
+  using index_type           = typename Extents::index_type;
+  using element_type         = std::remove_cv_t<ElementType>;
+  using owning_device_buffer = device_mdarray<element_type, Extents, LayoutPolicy, ContainerPolicy>;
+  using data_store           = std::variant<ElementType*, owning_device_buffer>;
+
+ public:
+  device_buffer(device_buffer const&) = delete;
+  device_buffer& operator=(device_buffer const&) = delete;
+
+  constexpr device_buffer(device_buffer&&) = default;
+  constexpr device_buffer& operator=(device_buffer&&) = default;
+
+  device_buffer(device_resources const& handle,
```

Comment:
To keep this consistent with our other APIs, can we add a factory function to create this? Something like `make_device_buffer` or `make_readonly_device_buffer` and `make_writeback_device_buffer` (or  have the latter 2 call the former?)